Refactor/consolidated inference improvements

SyedaAnshrahGillani opened the pull request:
This pull request introduces a series of significant improvements across the
  inference.py and prompting.py modules, enhancing the robustness, usability,
  and maintainability of the language model integration and prompt loading
  functionality.

  Key Features and Improvements:

   * Instance-Level Configuration for OllamaLanguageModel:
       * Allows users to configure inference parameters (e.g., seed,
         temperature, top_k, num_threads, num_ctx, max_output_tokens) at the
         OllamaLanguageModel instance level.
       * Introduces configurable request timeout for OllamaLanguageModel
         instances.
       * Improves usability by allowing reuse of configured instances without
         repeated parameter passing.
       * Establishes clear precedence for inference parameters: infer method
         arguments > instance-level configurations > method defaults.

   * Robust Parallel Gemini Inference:
       * Refactors GeminiLanguageModel.infer to utilize
         concurrent.futures.ThreadPoolExecutor.map.
       * Enhances robustness and "fail-fast" behavior by providing immediate
         exception propagation for parallel API calls.
       * Simplifies code and improves readability by removing manual future
         management.

   * Improved Prompt Template Validation:
       * Fixes an issue in read_prompt_template_structured_from_file to raise a
         ValueError on malformed or empty YAML/JSON files.
       * Ensures clear and actionable error messages, promoting "fail-fast"
         behavior for better debuggability.

   * Code Quality and Documentation:
       * Applied project-wide code formatting standards using isort and pyink to
         ensure consistency.
       * Enhanced documentation across inference.py, including expanded
         docstrings for BaseLanguageModel, OllamaLanguageModel,
         GeminiLanguageModel, OpenAILanguageModel, ScoredOutput, and
         InferenceOutputError to provide clearer explanations and better context.


  These consolidated improvements align with the project's goals of enhancing
  the language model integration and overall code health.


  Choose one: (Feature)

  How Has This Been Tested?

  The existing test suite was run to ensure that the refactoring and new
  features did not introduce any regressions.

   1 tox -e lint-src,lint-tests,py311

  Additionally, the ./autoformat.sh script was executed to apply the project's
  formatting standards. All existing tests pass successfully, and code
  formatting is compliant.

  Checklist:

  <!-- Please complete this checklist before submitting your PR. -->
   - [x] I have read and acknowledged Google's Open Source [Code of conduct 
     (https://opensource.google/conduct).
   - [x] I have read the [Contributing 
     (https://github.com/google/langextract/blob/main/CONTRIBUTING.md) page, and
     I either signed the Google Individual CLA 
     (https://cla.developers.google.com/about/google-individual) or am covered by
      my company's Corporate CLA 
     (https://cla.developers.google.com/about/google-corporate).
   - [ ] I have discussed my proposed solution with code owners in the linked
     issue(s) and we have agreed upon the general approach. (Please check this if
     you have, otherwise, it's understood that this is a direct contribution.)
   - [x] I have made any needed documentation changes, or noted in the linked
     issue(s) that documentation elsewhere needs updating.
   - [x] I have added tests, or I have ensured existing tests cover the changes.
   - [x] I have followed [Google's Python Style Guide 
     (https://google.github.io/styleguide/pyguide.html) and ran ./autoformat.sh
     over the affected code.

aksg87 commented:
Hi @SyedaAnshrahGillani, thanks for your contribution.

We recently added automated code formatting (#24). Please update your branch to incorporate these changes:

```bash
git fetch origin
git rebase origin/main
./autoformat.sh
git add -A
git commit -m "style: apply project formatting standards"
git push --force-with-lease
```

See our updated [CONTRIBUTING.md](https://github.com/google/langextract/blob/main/CONTRIBUTING.md) for details on:
- Running `./autoformat.sh` before submitting PRs
- Local testing with `tox -e format,lint-src,lint-tests`

Please also use the PR template and for larger changes please link to an issue or start a discussion to support the change.

SyedaAnshrahGillani commented:
Hi @aksg87, Thank you for the detailed feedback! I've addressed all the points you raised:

   * I've rebased the branch on origin/main and applied the project's formatting
     standards using ./autoformat.sh. The style changes are committed with the
     requested message.
   * The PR description has been updated to adhere to the new template, providing
     a comprehensive overview of the changes and their benefits.
   * I've also added further documentation to the inference.py module to improve
     clarity and context.

  The branch has been force-pushed with these updates. Please let me know if
  there's anything else needed for review.

aksg87 commented:
Thanks for these valuable improvements to the inference module. However, this PR needs to follow our [contributing guidelines](https://github.com/google/langextract/blob/main/CONTRIBUTING.md) and [PR template](https://github.com/google/langextract/blob/main/.github/PULL_REQUEST_TEMPLATE/pull_request_template.md) requirements:

**Missing Requirements:**
- **Linked Issues**: All non-trivial PRs must reference issues documenting motivation, context, and approach
- **Single Commit**: PRs should be squashed to one commit (currently 13 commits)  
- **Focused Scope**: Each PR should address one specific change - this covers multiple unrelated features (Ollama config, Gemini inference, prompt validation)
- **Passing Tests**: 2/4 checks are currently failing
- **Large Change Process**: For substantial multi-feature changes, create issues first for community discussion, then submit separate focused PRs

Per our guidelines, this PR will need to be closed to follow the proper process. Please create individual issues for each feature to allow community feedback, then submit focused single-commit PRs following our contribution requirements. The work is valuable and we'd welcome it resubmitted with proper structure.
